### PR TITLE
[SPARK-15864] [SQL] Fix Inconsistent Behaviors when Uncaching Non-cached Tables

### PR DESCRIPTION
**Pull request by gatorsmile**
#### What changes were proposed in this pull request?

To uncache a table, we have three different ways:
- _SQL interface_: `UNCACHE TABLE`
- _DataSet API_: `sparkSession.catalog.uncacheTable`
- _DataSet API_: `sparkSession.table(tableName).unpersist()`

When the table is not cached,
- _SQL interface_: `UNCACHE TABLE non-cachedTable` -> **no error message**
- _Dataset API_: `sparkSession.catalog.uncacheTable("non-cachedTable")` -> **report a strange error message:**
  `requirement failed: Table [a: int] is not cached`
- _Dataset API_: `sparkSession.table("non-cachedTable").unpersist()` -> **no error message**

This PR will make them consistent. No operation if the table has already been uncached.

In addition, this PR also removes `uncacheQuery` and renames `tryUncacheQuery` to `uncacheQuery`, and documents it that it's noop if the table has already been uncached
#### How was this patch tested?

Improved the existing test case for verifying the cases when the table has not been cached.
Also added test cases for verifying the cases when the table does not exist

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @rxin @hvanhovell @liancheng This is another issue related to `Cache` and `Uncache`. Actually, I am not sure if we should provide a SQL interface `UNCACHE TABLE IF EXISTS`. 

Thanks!

**Comment by liancheng** [comment]
From the perspective of better consistency, I'd prefer the current fix in the PR plus a new `UNCACHE TABLE ... IF EXISTS` syntax. However, this PR does introduce behavior breaking change.

@rxin What do you think?

(BTW, it's "Dataset" rather than "DataSet".)

**Comment by liancheng** [comment]
We do have a private method `tryUncacheQuery` in `CacheManager`, which exposes exactly the same semantics as `UNCACHE TABLE ... IF EXISTS`. But `Catalog` doesn't have an equivalent public method.

**Comment by rxin** [comment]
Hmmm maybe just throw an exception if the table does not exist, but no-op if the table is already uncached. I don't think we should change the behavior here.

**Comment by gatorsmile** [comment]
@rxin @liancheng I see. Since the existing Dataset API `sparkSession.catalog.uncacheTable("non-cachedTable")` issues an error if uncaching non-cached tables. Thus, to ensure both SQL statements and Dataset APIs have the same behavior. We still need to change one of them, right?

Will follow what @rxin said. No-op if the table is already uncached.

**Comment by gatorsmile** [comment]
@liancheng `tryUncacheQuery` and `uncacheQuery` are different. `tryUncacheQuery` does not unregister the accumulators, but `uncacheQuery` does it. This difference is also confusing.

`uncacheQuery`:
https://github.com/apache/spark/blob/468da03e23a01e02718608f05d778386cbb8416b/sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala#L112

`tryUncacheQuery `:
https://github.com/apache/spark/blob/468da03e23a01e02718608f05d778386cbb8416b/sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala#L127

I plan to unregister the accumulators in both APIs. Does that make sense? Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala [hunk 1]
```diff
@@ -323,7 +323,7 @@ class CatalogImpl(sparkSession: SparkSession) extends Catalog {
    * @since 2.0.0
    */
   override def uncacheTable(tableName: String): Unit = {
-    sparkSession.sharedState.cacheManager.uncacheQuery(sparkSession.table(tableName))
+    sparkSession.sharedState.cacheManager.tryUncacheQuery(query = sparkSession.table(tableName))
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala]: After this change, nobody is calling `sparkSession.sharedState.cacheManager.uncacheQuery`. Should we remove this API?

**Comment by cloud-fan** [comment]
hi @gatorsmile , do you wanna update it? now both `tryUncacheQuery` and `uncacheQuery` won't unregister accumulator

**Comment by gatorsmile** [comment]
@cloud-fan Thank you for letting me know it. Just updated the PR. Please let me know if anything needs a change. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala (hunk 1 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala]: maybe we can just make `uncacheTable` same with `tryUncacheQuery` and remove `tryUncacheQuery`

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala]: Sure, will do it. Thanks!

**Comment by cloud-fan** [comment]
LGTM, can you update the PR description?

cc @rxin for final sing-off as this is kind of a behaviour change.

**Comment by rxin** [comment]
Behavior change LGTM. 

@gatorsmile can you update the pr description? It is outdated.

**Comment by gatorsmile** [comment]
@cloud-fan @rxin @liancheng  Thank you for your reviews! 

The PR description is updated. Let me know if any change is needed. Thanks!

**Comment by cloud-fan** [comment]
> Dataset API: sparkSession.catalog.uncacheTable("non-cachedTable") -> report a strange error message: requirement failed: Table [a: int] is not cached
Dataset API: sparkSession.catalog.uncacheTable("non-cachedTable") -> no error message

hm?

**Comment by gatorsmile** [comment]
@cloud-fan Sorry, please refresh the browser. Just finishing the changes.

**Comment by cloud-fan** [comment]
thanks, merging to master/2.0!

**Pull request closed without merging**